### PR TITLE
plrust linting via plrustc custom driver

### PR DESCRIPTION
**Pull request by thomcc**

### CODE REVIEW

**Comment by thomcc** [comment]
I ended up significantly simplifying the build system. By which I mean, deleting – there's no need for a dedicated one any more, or at least not that one; 90% of what that one with multi-toolchain garbage is completely unnecessary now.

At the moment build is not so complex that someone can't build things by hand. I have a small script to do some checking and make sure the right flags are set, but it's not majore. I suspect we'll need to bring build system back (for an installer) but for now we're better without it.

---

In general this code is pretty messy (lots of commented out parts) and there's some real dumb nonsense I've done in it (I've just been hammering trying to get it working). Still, it completely works now and integration could start if you're brave. IOW, it seems to be able to compile stuff I throw at it -- including postgrestd and some pgx extensions. The lints also have tests... so I imagine it's somewhat ready.

### The summary of the build steps are: 
1. Checkout branch
2. Setup your rust toolchain to 1.67.1. Be sure to install rustc-dev component.
3. Run `./build.sh`, which will build plrustc and put a binary in `build/bin/plrustc`, which you can move to wherever you want.
4. To use set `RUSTC=/path/to/plrustc` in the env when running cargo. It behaves as a normal rustc by default. (E.g. All the lints are off by default, add `#![forbid(plrrust_lints)]` in the generated user crate to enable them).
5. Be sure to also build install postgrestd (from the 1.67.1 branch) as normal. You don't need to use plrustc to build it (any more), but it can't be the old install.

### Troubleshooting:

That works for me. But if you have issues with the above:

- Make sure that `path/to/plrustc -Vv` and `rustc -Vv` have identical output. Definitely make sure they're both claiming to be 1.67.1 stable.
- Try installing the `rust-src` and `llvm-tools-preview` components via rustup, although I think just rustc-dev should be needed.
- Ensure `--target $arch-postgres-linux-gnu` is passed when building the user crate.
- When building, try setting `PLRUSTC_SYSROOT` to the sysroot containing the postgrestd install. It's hard for me to imagine the case where this gets set to something different here, so this shouldn't be needed either.
- Try using a pgx version with https://github.com/tcdi/pgx/pull/1038 in it. I don't think this is needed any more, although we should probably take that commit eventually (doesn't need to be soon)

**Review by eeeebbbbrrrr — COMMENTED** [review]
Just making some notes here.  

I haven't even looked at the actual rust code yet.  Still going through the setup process.

File: plrustc/.cargo/config.toml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# [env]
+# # FIXME :(
```
[the remark this answers is not in the document]
Comment (thomcc) [reply plrustc/.cargo/config.toml:2]: Oh, I should uncomment it. This prevents you from needing to set `RUSTC_BOOTSTRAP=1` to run tests and such.

File: plrust/src/user_crate/crating.rs [hunk 2]
```diff
@@ -283,7 +283,7 @@ pub(crate) fn cargo_toml_template(crate_name: &str, version_feature: &str) -> to
         [profile.release]
         debug-assertions = true
         codegen-units = 1_usize
-        lto = "fat"
+        lto = "thin"
```
[the remark this answers is not in the document]
Comment (thomcc) [reply plrust/src/user_crate/crating.rs:286]: This is part of @workingjubilee's changes from #201, not mine.

File: build.sh [hunk 3]
```diff
@@ -0,0 +1,65 @@
+#!/usr/bin/env bash
```
[the remark this answers is not in the document]
Comment (thomcc) [reply build.sh:1]: I'll move it into the subdirectory.

File: build.sh [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+fi
+
+if [ "$SET_SYSROOT" = "1" ]; then
+    # Set fallback sysroot if requested.
+    export PLRUSTC_SYSROOT="$sysroot"
+fi
+
+echo "plrustc build starting..." >&2
+
+env RUSTFLAGS="$RUSTFLAGS" \
```
[the remark this answers is not in the document]
Comment (thomcc) [reply build.sh:54]: I'll look into detecting it.

File: plrust/src/user_crate/crating.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (workingjubilee) [reply plrust/src/user_crate/crating.rs:286]: Addressed in #201's latest commit.

**Review by eeeebbbbrrrr — APPROVED** [review]
I think this is good.  I was able to successfully use it to develop #207, so that's good.

**Comment by eeeebbbbrrrr** [comment]
@thomcc, feel free to merge soon as you're comfortable with this.

**Pull request merged**